pal/src/httpapi_sl.c

```diff
@@ -0,0 +1,492 @@
+                }
+
+                ret = BUFFER_content(responseContent,
+                        (const unsigned char **)&hname);
+                if (ret != 0) {
+                    LogError("Failed getting the response buffer content");
+                    ret = HTTPAPI_ALLOC_FAILED;
+                    goto contentDone;
+                }
+
+                memcpy(hname + offset, contentBuf, cnt);
+                offset += cnt;
+            }
+        } while (moreFlag);
+
+    contentDone:
+        if (ret < 0) {
+            BUFFER_unbuild(responseContent);
+            return ((HTTPAPI_RESULT)ret);
+        }
+    }
+
+    return (HTTPAPI_OK);
+}
+
+HTTPAPI_RESULT HTTPAPI_SetOption(HTTP_HANDLE handle, const char* optionName,
+        const void* value)
+{
+    HTTPAPI_RESULT result;
+    HTTPAPI_Object *apiH = (HTTPAPI_Object *)handle;
+    int len;
+
+    if ((apiH == NULL) ||
+            (optionName == NULL) ||
+            (value == NULL)) {
+        result = HTTPAPI_INVALID_ARG;
+    }
+    else if ((strcmp("x509EccCertificate", optionName) == 0) ||
+            (strcmp("x509certificate", optionName) == 0)) {
+        if (apiH->x509Certificate) {
+            free(apiH->x509Certificate);
+        }
+
+        len = strlen((char *)value);
+        apiH->x509Certificate = (char *)malloc((len + 1) * sizeof(char));
+        if (apiH->x509Certificate == NULL) {
+            result = HTTPAPI_ALLOC_FAILED;
+            LogInfo("unable to allocate memory for the client certificate"
+                    " in HTTPAPI_SetOption");
+        }
+        else {
+            strcpy(apiH->x509Certificate, (const char *)value);
+            result = HTTPAPI_OK;
+        }
+    }
+    else if ((strcmp("x509EccAliasKey", optionName) == 0) ||
+            (strcmp("x509privatekey", optionName) == 0)) {
+        if (apiH->x509PrivateKey) {
+            free(apiH->x509PrivateKey);
+        }
+
+        len = strlen((char *)value);
+        apiH->x509PrivateKey = (char *)malloc((len + 1) * sizeof(char));
+        if (apiH->x509PrivateKey == NULL) {
+            result = HTTPAPI_ALLOC_FAILED;
+            LogInfo("unable to allocate memory for the client private key"
+                    " in HTTPAPI_SetOption");
+        }
+        else {
+            strcpy(apiH->x509PrivateKey, (const char*)value);
+            result = HTTPAPI_OK;
+        }
+    }
+    else {
+        result = HTTPAPI_INVALID_ARG;
+        LogInfo("unknown option %s", optionName);
+    }
+
+    return (result);
+}
+
+HTTPAPI_RESULT HTTPAPI_CloneOption(const char* optionName, const void* value,
+        const void** savedValue)
+{
+    HTTPAPI_RESULT result;
+    size_t certLen;
+    char *tempCert;
+
+    if ((optionName == NULL) ||
+            (value == NULL) ||
+            (savedValue == NULL)) {
+        result = HTTPAPI_INVALID_ARG;
+    }
+    else if ((strcmp("x509EccCertificate", optionName) == 0) ||
+            (strcmp("x509certificate", optionName) == 0) ||
```

Comment:
Should use `OPTION_X509_ECC_CERT` and `SU_OPTION_X509_CERT` from `shared_util_options.h` instead of string literals